#!/bin/bash
#This script will take ATAC-seq Peaks in B, CD4, CD8, Lsk, and NK cells and take the union of all peaks (unionbedg) into one combined file
#That file is then split back into separate B, CD4, CD8, Lsk, and NK files using an R script
#The purpose of this is to expand the ATAC-seq peaks in each cell to encompass the ATAC-seq peaks of all the other cells combined
#The tag coverage is then generated for each ATAC-seq peak for each cell and then an R script is used to generate unique ATAC-seq peaks
#The tags are normalized by an R script, and the unique ATAC seq peaks and a heatmap are generated by another R script

infodir=/mnt/data1/John/Pioneer-Factors/info
maindir=/mnt/data1/John/Pioneer_Factors
scriptdir=/mnt/data1/John/Pioneer-Factors
filedir=/mnt/data1/John/Pioneer-Factors/sample_files
paralleldir=/mnt/data1/John/Pioneer-Factors/parallel_commands
r_dir=/mnt/data1/John/Pioneer-Factors/R_Scripts
datadir=/mnt/data1/VahediLab/PTF_Team/Data
dir0=$maindir

mkdir -p $dir0/Analysis/ATAC_seq/Union/R_output_merge_coverage/
cd $dir0/Analysis/ATAC_seq/Union/R_output_merge_coverage/
cp $dir0/Analysis/ATAC_seq/Union/R_output/B_CD4_CD8_CMP_GMP_GN_Lsk_MEP_Mono_NK_union_log2_merge.bed B_union_log2_merge.bed
cp $dir0/Analysis/ATAC_seq/Union/R_output/B_CD4_CD8_CMP_GMP_GN_Lsk_MEP_Mono_NK_union_log2_merge.bed CD4_union_log2_merge.bed
cp $dir0/Analysis/ATAC_seq/Union/R_output/B_CD4_CD8_CMP_GMP_GN_Lsk_MEP_Mono_NK_union_log2_merge.bed CD8_union_log2_merge.bed
cp $dir0/Analysis/ATAC_seq/Union/R_output/B_CD4_CD8_CMP_GMP_GN_Lsk_MEP_Mono_NK_union_log2_merge.bed CMP_union_log2_merge.bed
cp $dir0/Analysis/ATAC_seq/Union/R_output/B_CD4_CD8_CMP_GMP_GN_Lsk_MEP_Mono_NK_union_log2_merge.bed GMP_union_log2_merge.bed
cp $dir0/Analysis/ATAC_seq/Union/R_output/B_CD4_CD8_CMP_GMP_GN_Lsk_MEP_Mono_NK_union_log2_merge.bed GN_union_log2_merge.bed
cp $dir0/Analysis/ATAC_seq/Union/R_output/B_CD4_CD8_CMP_GMP_GN_Lsk_MEP_Mono_NK_union_log2_merge.bed Lsk_union_log2_merge.bed
cp $dir0/Analysis/ATAC_seq/Union/R_output/B_CD4_CD8_CMP_GMP_GN_Lsk_MEP_Mono_NK_union_log2_merge.bed MEP_union_log2_merge.bed
cp $dir0/Analysis/ATAC_seq/Union/R_output/B_CD4_CD8_CMP_GMP_GN_Lsk_MEP_Mono_NK_union_log2_merge.bed Mono_union_log2_merge.bed
cp $dir0/Analysis/ATAC_seq/Union/R_output/B_CD4_CD8_CMP_GMP_GN_Lsk_MEP_Mono_NK_union_log2_merge.bed NK_union_log2_merge.bed

#------------------BEDTOOLS COVERAGE------------------------------------------------------
#-------Sorting Union BED Files to BAM File Order and generated parallel commands---------
rm -f $paralleldir/ATAC_coverage_parallel.txt
for i in `ls *union_log2_merge.bed`; 
  do
    filename=${i%*_union_log2_merge.bed}
    sed -e 's/chr1	/chr1A	/' $i | sort -k1,1 -k2,2n | sed -e 's/chr1A/chr1/' > ${filename}_union_merge_split_sorted.bed
    datafile=`find /mnt/data1/VahediLab/PTF_Team/Data/ATAC_seq/bam | grep "$filename" | grep -v "bai"`	
    echo "${dir0}/Analysis/ATAC_seq/Union/R_output_merge_coverage/${filename}_union_merge_split_sorted_coverage.bedgraph ${dir0}/Analysis/ATAC_seq/Union/R_output_merge_coverage/${filename}_union_merge_split_sorted.bed ${datafile}"	
  done >> $paralleldir/ATAC_coverage_parallel.txt	
#if [ ! -e ${dir0}/Analysis/ATAC_seq/B_bam_count.bedgraph ]; then
parallel --xapply --dryrun --colsep ' ' -a $paralleldir/ATAC_coverage_parallel.txt "bedtools coverage -sorted -g $infodir/mm10.genome -counts -a {2} -b {3} > {1}"	
parallel --xapply --colsep ' ' -a $paralleldir/ATAC_coverage_parallel.txt "bedtools coverage -sorted -g $infodir/mm10.genome -counts -a {2} -b {3} > {1}"
#fi
#for i in `ls *_union_merge_split_sorted_coverage.bedgraph`; 
#  do
#    filename=${i%*_union_log2_merge.bed}
#    filename=`echo "$i" | cut -d'.' -f1`
#    awk 'BEGIN{OFS="\t"} {print $1,$2,$3,$5}' ${i} > ${filename}_cond.bedgraph	##Removes the old tag counts, keeping only the new tag counts
#  done
#filename="$(ls *bam_count.bedgraph | cut -d'_' -f1 | paste -sd "_" -)"	##Stores a combined filename
#bedtools unionbedg -i `ls *bam_count_cond.bedgraph` | sort -k1,1 -k2,2n | bedtools merge -i stdin -c 4,5,6,7,8 -o sum,sum,sum,sum,sum > ${filename}_bam_count_all.bedgraph

#---------------------R Script that gives ATAC UNIQUE PEAKS and converting to BEDGRAPH----
#Rscript --verbose ${r_dir}/Normalize_ATAC_Peaks.R $dir0	
#Rscript --verbose ${r_dir}/Heatmap1.R $dir0	